rpc/handlers_test.go

```diff
@@ -239,132 +399,222 @@ func TestGetTransactionByHash(t *testing.T) {
+		return mainnetGw.Transaction(context.Background(), hash)
+	}).Times(len(tests))
+
 	for name, test := range tests {
 		t.Run(name, func(t *testing.T) {
 			hash, err := new(felt.Felt).SetString(test.hash)
 			require.NoError(t, err)
-			res, rpcErr := handler.GetTransactionByHash(hash)
-			require.Nil(t, rpcErr)
 
 			expectedMap := make(map[string]any)
 			require.NoError(t, json.Unmarshal([]byte(test.expected), &expectedMap))
 
+			res, rpcErr := handler.GetTransactionByHash(hash)
+			require.Nil(t, rpcErr)
+
 			resJson, err := json.Marshal(res)
 			require.NoError(t, err)
 			resMap := make(map[string]any)
 			require.NoError(t, json.Unmarshal(resJson, &resMap))
+
 			assert.Equal(t, expectedMap, resMap, string(resJson))
 		})
 	}
 }
+
+func TestGetTransactionReceiptByHash(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	mockReader := mocks.NewMockReader(mockCtrl)
+	handler := rpc.New(mockReader, utils.MAINNET)
+
+	t.Run("transaction not found", func(t *testing.T) {
+		txHash := new(felt.Felt).SetBytes([]byte("random hash"))
+		mockReader.EXPECT().GetTransactionByHash(txHash).Return(nil, errors.New("tx not found"))
+
+		tx, rpcErr := handler.GetTransactionReceiptByHash(txHash)
+		assert.Nil(t, tx)
+		assert.Equal(t, rpc.ErrTxnHashNotFound, rpcErr)
+	})
+
+	mainnetGw, closer := testsource.NewTestGateway(utils.MAINNET)
+	defer closer()
+
+	block0, err := mainnetGw.BlockByNumber(context.Background(), 0)
+	require.NoError(t, err)
+
+	tests := map[string]struct {
+		index    int
+		expected string
+	}{
+		"with contract addr": {
+			index: 0,
+			expected: `{
+					"type": "DEPLOY",
+					"transaction_hash": "0xe0a2e45a80bb827967e096bcf58874f6c01c191e0a0530624cba66a508ae75",
+					"actual_fee": "0x0",
+					"status": "ACCEPTED_ON_L2",
+					"block_hash": "0x47c3637b57c2b079b93c61539950c17e868a28f46cdef28f88521067f21e943",
+					"block_number": 0,
+					"messages_sent": [],
+					"events": [],
+					"contract_address": "0x20cfa74ee3564b4cd5435cdace0f9c4d43b939620e4a0bb5076105df0a626c6"
+				}`,
+		},
+		"without contract addr": {
+			index: 2,
+			expected: `{
+					"type": "INVOKE",
+					"transaction_hash": "0xce54bbc5647e1c1ea4276c01a708523f740db0ff5474c77734f73beec2624",
+					"actual_fee": "0x0",
+					"status": "ACCEPTED_ON_L2",
+					"block_hash": "0x47c3637b57c2b079b93c61539950c17e868a28f46cdef28f88521067f21e943",
+					"block_number": 0,
+					"messages_sent": [
+						{
+							"to_address": "0xc84dd7fd43a7defb5b7a15c4fbbe11cbba6db1ba",
+							"payload": [
+								"0xc",
+								"0x22"
+							]
+						}
+					],
+					"events": []
+				}`,
+		},
+	}
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			txHash := block0.Transactions[test.index].Hash()
+			mockReader.EXPECT().GetTransactionByHash(txHash).Return(block0.Transactions[test.index], nil)
+			mockReader.EXPECT().GetReceipt(txHash).Return(block0.Receipts[test.index], block0.Hash, block0.Number, nil)
```

Comment:
We could probably consider this https://gist.github.com/thiagozs/4276432d12c2e5b152ea15b3f8b0012e#asserting-call-order

```suggestion
			callFirst := mockReader.EXPECT().GetTransactionByHash(txHash).Return(block0.Transactions[test.index], nil)
			mockReader.EXPECT().GetReceipt(txHash).Return(block0.Receipts[test.index], block0.Hash, block0.Number, nil).After(callFirst)
```

Reply:
No, we should not because it increases coupling to our implementation thus making our test fragile. We are using mock as data providers as long as data is being provided as expected we shouldn't assert the order